### PR TITLE
Refactor LocalWaker

### PR DESCRIPTION
**Pull request by botika**
~Simplify LocalWaker register method with `replace` mutable pointer method~

Refactor `LocalWaker`: `is_registered` has no real use because the only use of the registry is to wake up the task it refers to. But `wake` does not need the registration status, the behavior is the same whether a task is registered or not. So it has been removed in favor of `wake`. Without this method the use of `UnsafeCell` is not necessary so I have proposed a later refactor from `UnsafeCell` to `Cell`.

### CODE REVIEW

**Comment by botika** [comment]
It seems there are clippy bugs in the current version. Now I upload the fix.

File: actix-server/src/signals.rs [hunk 1]
```diff
@@ -1,8 +1,10 @@
 use std::future::Future;
-use std::io;
 use std::pin::Pin;
 use std::task::{Context, Poll};
 
+#[cfg(not(unix))]
+use std::io;
```
Comment (robjtede) [inline actix-server/src/signals.rs:6]: inline this since it's only used in one place under the cfg attr

File: actix-utils/src/task.rs [hunk 2]
```diff
@@ -45,12 +45,7 @@ impl LocalWaker {
     ///
     /// Returns `true` if waker was registered before.
     pub fn register(&self, waker: &Waker) -> bool {
-        unsafe {
-            let w = self.waker.get();
-            let is_registered = (*w).is_some();
-            *w = Some(waker.clone());
-            is_registered
-        }
+        unsafe { self.waker.get().replace(Some(waker.clone())).is_some() }
```
Comment (robjtede) [inline actix-utils/src/task.rs:48]: Binding names are good.

```rust
let w = self.waker.get();
let last_waker = w.replace(Some(waker.clone()));
// indicates was registered
last_waker.is_some()
```

**Review by robjtede — CHANGES_REQUESTED** [review]

**Comment by fakeshadow** [comment]
We could just use a Cell though. The counter's waker is rarely used and when it's used the system could already be bound to CPU.

**Comment by botika** [comment]
> We could just use a Cell though. The counter's waker is rarely used and when it's used the system could already be bound to CPU.

`Waker` not implement `Copy`

**Comment by fakeshadow** [comment]
https://github.com/fakeshadow/actix-net/blob/2d35038600cc74b527f6e84e5dcf50e27ed48c48/actix-rt/src/net.rs#L106
This is the counter I use in a fork.

**Comment by botika** [comment]
https://doc.rust-lang.org/std/cell/struct.Cell.html#impl-1
Well, I guess you won't use the get method at all.

**Comment by fakeshadow** [comment]
Yea we don't use that. `is_registered` is an API I can't find usage internally. Even the register API just replace and check the old one as option

**Comment by botika** [comment]
The uses of LocalWaker `take` in `actix-web`:
- https://github.com/actix/actix-web/blob/b5812b15f07271850367b085e4933a0989dd84f9/actix-http/src/h1/payload.rs#L185
- https://github.com/actix/actix-web/blob/b5812b15f07271850367b085e4933a0989dd84f9/actix-multipart/src/server.rs#L728-L730
Can substitute by `wake`. I don't think I left any.

I `is_registered` is not used in `actix-web`

**Comment by robjtede** [comment]
I wouldn't create breaking change without good reason at this point. Keep this as a refactor.

**Comment by fakeshadow** [comment]
I think it's OK to remove `is_registered`. There is no instance of using it in any of actix repo and I doubt anyone using it by importing actix-utils for a waker type.

Besides that `LocalWaker` doesn't expose an API for comparing the waker so you know it's registered but you don't know if you still have the same context by comparing the waker.(Which could have false positive but I'm not sure).
I feel it's safer to always replace the waker than check it.

**Comment by robjtede** [comment]
Changing public API is always a breaking change. Please keep it the same for this PR. Adding deprecation attribute to `is_registered` with a note about its lack of usefulness would be acceptable.

**Comment by botika** [comment]
> Changing public API is always a breaking change. Please keep it the same for this PR. Adding deprecation attribute to `is_registered` with a note about its lack of usefulness would be acceptable.

Totally agree. Sorry now I put this methods deprecated. I do not know what I was thinking of.

File: actix-server/src/signals.rs (hunk 1 above)
Comment (robjtede) [reply actix-server/src/signals.rs:6]: this has come back

Comment (botika) [reply actix-server/src/signals.rs]: Sorry :parachute:  Now it's fixed

**Review by robjtede — APPROVED** [review]

**Review by fakeshadow — APPROVED** [review]

**Pull request merged**